### PR TITLE
L3 performance and scaling enhancements HLD initial version

### PR DESCRIPTION
**Pull request by Arvind-Brcm**
L3 performance and scaling enhancements HLD initial version

### CODE REVIEW

File: doc/L3_performance_and_scaling_enchancements_HLD.md [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+  Fpmsyncd listens to Netlink messages for Route add/delete messages and updates the APP_DB.
+  Current behaviour in Fpmsyncd
+   - When fpmsyncd get a route first it tries to get the master device name from the rt_table attribute of route object.
+     This is done to check if the route belongs to VNET_ROUTE_TABLE.
+  - To get a Master device name, fpmsyncd does a lookup in its local link cache. 
+  - If the lookup in the local link cache fails, fpmsyncd updates cache by getting the configured links from the kernel.
+  
+  The problem here is if there are no VNET present on the system, this lookup will always fail and cache is updated for every route.
+  This seems to slow down the rate which route programed for the global route table.
```
Comment (lguohan) [inline doc/L3_performance_and_scaling_enchancements_HLD.md]: @prsunny , can you take a look at this problem?

File: doc/L3_performance_and_scaling_enchancements_HLD.md [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+To measure route programming time, BGP routes were advertised to a SONiC router and timed how long it took for these routes to be installed in the ASIC.
+
+### Table 2: IPv4 prefix route programming time
+
+| Routes                  | time taken on AS7712(Tomahawk) |
+| ----------------------- | ------------------------------ |
+| 10k IPv4 prefix routes  | 11 seconds                     |
+| 30k IPv4 prefix routes  | 30 seconds                     |
+| 60k IPv4 prefix routes | 48 seconds                     |
+| 90k IPv4 prefix routes  | 68 seconds                     |
```
Comment (madhupalu) [inline doc/L3_performance_and_scaling_enchancements_HLD.md]: Do you have breakdown of programming time of each layer ex: orchagent or fpmsyncd/syncd?

File: doc/L3_performance_and_scaling_enchancements_HLD.md [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+
+
+
+### Table 3: IPv6 prefix route programming time
+
+| Routes                             | time taken on AS7712(Tomahawk) |
+| ---------------------------------- | -----------------------------  |
+| 10k IPv6 route with prefixes > 64b | 11 seconds                     |
+| 30k IPv6 route with prefixes >64b  | 30 seconds                     |
```
Comment (madhupalu) [inline doc/L3_performance_and_scaling_enchancements_HLD.md]: please list out all the usecases will need perf enhancements?
Ex: 1) Sonic boot time 2) graceful restart 3) download routes etc.

File: doc/L3_performance_and_scaling_enchancements_HLD.md [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+  The meta_sai layer in sairedis iterates over this route list and creates the meta objects for every route but only one Redis DB message will be generated for the route list. 
+  Therefore using the sairedis bulk APIs reduces the number of Redis messages. 
+  
+  The ASIC doesn't support bulk route creation/deletion,so syncd still processes one route at time and updates the ASIC.
+
+  So, the saving achieved by using bulk APIs will be number of Redis message generated.
+
+  Bulking of Route updates will be enabled in Orchagent. Orchagent will bulk 64 updates and send to Sairedis. 
+  A new timer will be introduced in orchagent to flush the outstanding updates every second.
+
```
Comment (madhupalu) [inline doc/L3_performance_and_scaling_enchancements_HLD.md]: Bulking idea is nice but It introduces the delay for non bulk cases too. Can this timer tuned(starts off with low value and grow exponenetially, simiar to tcp)?

File: doc/L3_performance_and_scaling_enchancements_HLD.md [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+## 5.3 Regresssion Testcases
+| Testcase number | Testcase                                                    | Result | Time taken |
+| --------------- | ----------------------------------------------------------- | ------ | ---------- |
+|              1. | Measure the convergence time with link flaps                |        |            |
+|              2. | Measure the convergence  time with Link flaps on ECMP paths |        |            |
+|              3. | Clear bgp neighbors to check all routes and forwarding      |        |            |
+|              4. | Clear neigh table and check all routes and forwarding       |        |            |
+|              5. | Clear mac table and check all routes and forwarding         |        |            |
+|              6. | Test across warm reboot , Orchagt/Syncd restart and upgrade |        |            |
+
```
Comment (madhupalu) [inline doc/L3_performance_and_scaling_enchancements_HLD.md]: Add these tests with warm reboot/graceful restart?

File: doc/L3_performance_and_scaling_enchancements_HLD.md [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+  By using Sairedis bulk APIs, orchagent will call these APIs with a list of routes and their attributes.
+  The meta_sai layer in sairedis iterates over this route list and creates the meta objects for every route but only one Redis DB message will be generated for the route list. 
+  Therefore using the sairedis bulk APIs reduces the number of Redis messages. 
+  
+  The ASIC doesn't support bulk route creation/deletion,so syncd still processes one route at time and updates the ASIC.
+
+  So, the saving achieved by using bulk APIs will be number of Redis message generated.
+
+  Bulking of Route updates will be enabled in Orchagent. Orchagent will bulk 64 updates and send to Sairedis.
```
Comment (stcheng) [inline doc/L3_performance_and_scaling_enchancements_HLD.md]: any reason why the bulk size is 64?

File: doc/L3_performance_and_scaling_enchancements_HLD.md (hunk 1 above)
Comment (prsunny) [reply doc/L3_performance_and_scaling_enchancements_HLD.md]: This is fixed as part of the PR https://github.com/Azure/sonic-swss/pull/1048

**Pull request merged**